File: day2/solution.py
with open("input.txt") as f: print((wins := {"A Y", "B Z", "C X"},ties := {"A X", "B Y", "C Z"},to_win := {"A":"Y", "B":"Z", "C":"X"},to_tie := {"A":"X", "B":"Y", "C":"Z"},to_lose := {"A":"Z", "B":"X", "C":"Y"},values := {"X":1, "Y":2, "Z":3},matches := f.read().split("\n"),p1:=sum([6*(match in wins) + 3*(match in ties) + values[match[-1]] if match!="" else 0 for match in matches]),p2:=sum([6*(match[-1] == "Z")+values[to_win[match[0]]]*(match[-1] == "Z") + 3*(match[-1] == "Y")+values[to_tie[match[0]]]*(match[-1] == "Y") + (match[-1] == "X")*values[to_lose[match[0]]] if match!="" else 0 for match in matches]))[-2:])

File: day2/solution.cpp
#include <set>
#include <map>

#include "../common/filereader.h"

int main() {

    auto matches = aoc_read_lines_raw();

    std::map<char, int> values = {{'X', 1}, {'Y', 2}, {'Z', 3}};

    // P1
    std::set<std::string> wins = {"A Y", "B Z", "C X"};
    std::set<std::string> ties = {"A X", "B Y", "C Z"};
    std::set<std::string> loses = {"A Z", "B X", "C Y"};
    
    size_t score = 0;
    for (auto& match : matches) {
        if (wins.find(match) != wins.end()) score += 6;
        if (ties.find(match) != ties.end()) score += 3;
        score += values[match.back()];
    }
    std::cout << score << std::endl;

    // P2
    std::map<char, char> to_win = {{'A', 'Y'}, {'B', 'Z'}, {'C', 'X'}};
    std::map<char, char> to_tie = {{'A', 'X'}, {'B', 'Y'}, {'C', 'Z'}};
    std::map<char, char> to_lose = {{'A', 'Z'}, {'B', 'X'}, {'C', 'Y'}};

    score = 0;
    for (auto& match : matches) {
        if (match.back() == 'Z') score += 6 + values[to_win[match.front()]];
        if (match.back() == 'Y') score += 3 + values[to_tie[match.front()]];
        if (match.back() == 'X') score += values[to_lose[match.front()]];
    }
    std::cout << score << std::endl;

    return 0;
}